### PR TITLE
Add Hvec space to de Rham sequence, if requested

### PR DESCRIPTION
**Pull request by vcarlier**
The "Hvec" space is the (H^1)^d space taken as the product of d identical Vh^0 spaces.

List of modifications
---------------------
- Add the "Hvec" space to `DiscreteDerham` objects, if requested to the `discretize_derham` function through the optional argument `get_H1vec_space` (default is `False`).
- Add a geometric projection for this space, which is a component-wise interpolation combined with the pull_back for vector fields.
- Add documentation to the `discretize_derham` function and the `DiscreteDerham` class.
- Add tests for the subclasses of `GlobalProjector`.

### CODE REVIEW

File: psydac/api/feec.py [hunk 1]
```diff
@@ -83,6 +92,11 @@ def V2(self):
     def V3(self):
         return self._spaces[3]
 
+    @property
+    def Vvec(self):
+        assert self.has_vec
```
Comment (spossann) [inline psydac/api/feec.py:97]: docstring missing. What space is this?

File: psydac/feec/basis_projectors.py [hunk 2]
```diff
@@ -0,0 +1,717 @@
...
+from psydac.fem.basic import FemField
+from psydac.utilities.utils import roll_edges
+
+from copy import deepcopy
+
+
+
+class BasisProjectionOperator(LinearOperator):
+    """
+    Class for "basis projection operators" PI_ijk(fun Lambda_mno) in the general form BP * P * DOF * EV^T * BV^T.
```
Comment (spossann) [inline psydac/feec/basis_projectors.py:22]: `BP * P * DOF * EV^T * BV^T` is not really applicable yet in Psydac, is it?

File: psydac/feec/global_projectors.py [hunk 3]
```diff
@@ -673,6 +735,24 @@ def evaluate_dofs_2d_2form(
                     F[i1, i2] += quad_w1[i1, g1] * quad_w2[i2, g2] * \
                             f(quad_x1[i1, g1], quad_x2[i2, g2])
 
+#------------------------------------------------------------------------------    
+def evaluate_dofs_2d_vec(
```
Comment (spossann) [inline psydac/feec/global_projectors.py:739]: Why not call the `dof_kernels` here?

**Review by spossann — APPROVED** [review]

File: psydac/api/discretization.py [hunk 4]
```diff
@@ -81,18 +81,42 @@ def change_dtype(V, dtype):
...
+    Parameters
+    ----------
+
+    derham : sympde.topology.space.Derham
+        The symbolic Derham sequence
+
+    domain_h   : Geometry
+        Discrete domain where the spaces will be discretized
+        
+    get_vec : Bool
```
Comment (campospinto) [inline psydac/api/discretization.py:98]: to be consistent with the notation used elsewhere, I would call this argument `get_H1vec_space`

File: psydac/feec/pull_push.py [hunk 5]
```diff
@@ -64,6 +65,36 @@ def f_logical(eta1):
 #==============================================================================
 # 2D PULL-BACKS
 #==============================================================================
+def pull_2d_v(f, F):
```
Comment (campospinto) [inline psydac/feec/pull_push.py:68]: I would prefer a more explicit name like `pull_2d_vec`

File: psydac/feec/pull_push.py [hunk 6]
```diff
@@ -161,29 +192,47 @@ def f_logical(eta1, eta2):
 # TODO [YG 05.10.2022]:
 # Remove? But it makes sense to return a vector-valued function...
 
-def pull_3d_v(funcs_ini, mapping):
+def pull_3d_v(f, F):
```
Comment (campospinto) [inline psydac/feec/pull_push.py:195]: I would prefer a more explicit name like `pull_3d_vec`

File: psydac/feec/pull_push.py [hunk 7]
```diff
@@ -64,6 +65,36 @@ def f_logical(eta1):
 #==============================================================================
 # 2D PULL-BACKS
 #==============================================================================
+def pull_2d_v(f, F):
+    #We should check if the metric terms are really the good ones!
```
Comment (campospinto) [inline psydac/feec/pull_push.py:69]: So, did you check ?

File: psydac/feec/global_projectors.py [hunk 8]
```diff
@@ -147,6 +149,9 @@ def __init__(self, space, nquads = None):
                     if quad_x[j] is None:
                         u, w = uw[j]
                         global_quad_x, global_quad_w = quadrature_grid(V.histopolation_grid, u, w)
+                        #"roll" back points to the interval to ensure that the quadrature points are
+                        #in the domain. Probably only usefull on periodic cases
+                        roll_edges(V.domain, global_quad_x)
```
Comment (campospinto) [inline psydac/feec/global_projectors.py:154]: is this related to the H1vec space, or fixing a pre-existing bug ?

File: psydac/feec/global_projectors.py [hunk 9]
```diff
@@ -564,6 +569,63 @@ def __call__(self, fun):
...
+    """
+    Projector from H1^3 = H1 x H1 x H1 to a conforming finite element space, i.e.
+    a finite dimensional subspace of H1^3, constructed with tensor-product
+    B-splines in 2 or 3 dimensions.
+    This is a global projector constructed over a tensor-product grid in the
+    logical domain. The vertices of this grid are obtained as the tensor
+    product of the 1D splines' Greville points along each direction.
+    Parameters
+    ----------
+    H1vec : ProductFemSpace
```
Comment (campospinto) [inline psydac/feec/global_projectors.py:582]: isn't the argument of the GlobalProjector constructor called `space` ?

File: psydac/api/feec.py [hunk 10]
```diff
@@ -4,25 +4,47 @@
...
+    
+    Parameters
+    ----------
+
+    mapping : Mapping
+        The mapping from the logical space to the physical space of the discrete De Rham.
+
+    get_vec : Bool
+        True to also get the "Hvec" space discretizing (H1)^n vector fields
+
```
Comment (yguclu) [inline psydac/api/feec.py:28]: I find it strange that this flag should be called `get_vec`, given that the H1-vector space is passed as an input to the constructor of this class.

**Review by yguclu — CHANGES_REQUESTED** [review]
Looks good, but I have a few comments.

**Review by vcarlier — COMMENTED** [review]
Am I reviewing when trying to reply?

File: psydac/api/ast/utilities.py [hunk 11]
```diff
@@ -386,23 +386,23 @@ def compute_atoms_expr_mapping(atomic_exprs, indices_quad,
...
+        # ...
 
-        # ...
+        # ...
         val_name = SymbolicExpr(atom).name + '_values'
         val      = IndexedBase(val_name)[indices_quad]
         expr     = IndexedBase(element_name)[idxs] * test_fun
         updates += [AugAssign(val, '+', expr)]
-        # ...
+        # ...
```
[the remark this answers is not in the document]
Comment (vcarlier) [reply psydac/api/ast/utilities.py:405]: This is probably a bug

File: psydac/api/ast/utilities.py [hunk 12]
```diff
@@ -331,14 +331,14 @@ def compute_atoms_expr_field(atomic_exprs, indices_quad,
...
 
-        # ...
+        # ...
         args     = [IndexedBase(field_name)[idxs], test_fun]
         val_name = SymbolicExpr(atom).name + '_values'
         val      = IndexedBase(val_name)[indices_quad]
         updates += [AugAssign(val,'+',Mul(*args))]
-        # ...
+        # ...
```
[the remark this answers is not in the document]
Comment (vcarlier) [reply psydac/api/ast/utilities.py:341]: This is probably a git bug

File: psydac/api/feec.py [hunk 13]
```diff
@@ -4,25 +4,46 @@
...
     """ Represent the discrete De Rham sequence.
+    Should be initialized via discretize_derham function in api.discretization.py
+    
+    Parameters
+    ----------
+
+    mapping : Mapping
+        The mapping from the logical space to the physical space of the discrete De Rham.
+        
+    *spaces : list of
```
Comment (yguclu) [inline psydac/api/feec.py:28]: List of `FemSpace` objects?

File: psydac/api/feec.py (hunk 1 above)
Comment (yguclu) [reply psydac/api/feec.py:109]: Yes, I think we should add docstrings for all of these properties. We should have done it before, could you please write one-line docstrings? Please let me know if you have any questions.

File: psydac/feec/global_projectors.py [hunk 14]
```diff
@@ -791,3 +878,32 @@ def evaluate_dofs_3d_3form(
...
+#------------------------------------------------------------------------------
+def evaluate_dofs_3d_vec(
+        intp_x1, intp_x2, intp_x3, # interpolation points
+        F1, F2, F3,                # array of degrees of freedom (intent out)
+        f1, f2, f3,                # input scalar function (callable)
+        ):
+    
+    # evaluate input functions at interpolation points (make sure that points are in [0, 1])
+    
+
```
Comment (yguclu) [inline psydac/feec/global_projectors.py:891]: Usually two consecutive empty lines are only used between separate functions or classes. These can be removed here completely.

**Review by yguclu — CHANGES_REQUESTED** [review]

File: psydac/api/feec.py (hunk 10 above)
Comment (yguclu) [reply psydac/api/feec.py]: Checking the last space is an elegant solution. :+1:

File: psydac/feec/global_projectors.py (hunk 8 above)
Comment (yguclu) [reply psydac/feec/global_projectors.py:154]: Can you please further explain what is happening here? `roll_edges` is only relevant in the periodic case, why do you think it is necessary now?

File: psydac/feec/pull_push.py (hunk 5 above)
Comment (vcarlier) [reply psydac/feec/pull_push.py]: h1vec might be the best choice cause maybe someday someone will need an l2vec space

File: psydac/feec/global_projectors.py (hunk 8 above)
Comment (yguclu) [reply psydac/feec/global_projectors.py:154]: @vcarlier Have you answered my question somewhere else?

**Review by yguclu — CHANGES_REQUESTED** [review]

File: psydac/feec/global_projectors.py (hunk 8 above)
Comment (vcarlier) [reply psydac/feec/global_projectors.py:156]: Not yet I was doing tests to be sure that you need it and this is indeed the case : 
In the non periodic case roll edges does nothing so you don't care applying it all the time.
In the periodic case, if you don't use it you might have quadrature point outside of the domain, which might cause problem when you want to define your function by it's value on the domain but it's not periodic on R (for example `x*(x-1)` define a continuous function when restricted to (0,1) with periodic BC, but it's not a periodic function on R so if you don't roll back your going to evaluate it outside of your domain and have wrong values, see `test_derham_projector_2d_hdiv_2` in `test_global_projectors.py` )

File: psydac/feec/tests/test_global_projectors.py [hunk 15]
```diff
@@ -83,6 +87,146 @@ def test_L2_projector_1d(domain, ncells, degree, periodic, nquads):
...
+    # Test if max-norm of error is <= TOL
+    maxnorm_error = abs(vals_u0 - vals_f).max()
+    print(ncells, maxnorm_error)
+    maxnorm_error = abs(vals_u1_1 - vals_f).max()
+    print(ncells, maxnorm_error)
+    maxnorm_error = abs(vals_u2 - vals_f).max()
+    print(ncells, maxnorm_error)
+    maxnorm_error = abs(vals_ux_1 - vals_f).max()
+    print(ncells, maxnorm_error)
+#    assert maxnorm_error <= 1e-14
```
[the remark this answers is not in the document]
Comment (vcarlier) [reply psydac/feec/tests/test_global_projectors.py]: I can just assert that the error is relatively small with a ad hoc defined threshold, if a future PR break the projections then this tests will fail.

**Review by yguclu — APPROVED** [review]
Good job!

**Pull request merged**